### ISSUE TITLE
[Feature Request] Mark index as non essential

### ISSUE THREAD

**Issue by gary-harpaz** [elastic/elasticsearch#41367]
Hi,

So in our production cluster we have important buisness data. All these indices have 1 or more replica shards.

Alongside this data we have non essential log data output by various applications.
We defined these non essential indices with 0 replicas to save space.

During times of instability a node crash containing such a non essential shard cause a status red non recoverable error in the cluster.
Panic throughout the devops team, alerts firing as you can imagine. 
Really the behavior I would want in this scenario is to just begin a new index.
Currently our solution is to manually delete these indices when this happens.
All this because of some lost log data.

Thanks,

Gary

**Comment by martijnvg** [comment]
Assuming that you use the cluster health api to determine the status, maybe in you case use the index filter option to determine the status of specific indices? So GET `/_cluster/health/production-indices*` instead of GET `/_cluster/health`? This way if there are problems with less important indices, then no alarms will go off.

**Comment by gary-harpaz** [comment]
I'm not sure this is just monitoring issue.
Because this is non recoverable (primary shard disappeard) I think it blocks recovery of other shards that do have replicas.

**Comment by martijnvg** [comment]
Are you observing that yellow indices are not recovering their replica shards? 

If an index is red then it should not block other indices that are yellow from recovering their replica shards. The global cluster status in your case will remain to be red, because for certain indices there are no primary shards for recovery to start recovery. So deleting these indices will change the global status from red to green.

**Comment by gary-harpaz** [comment]
The node that crashed contained a lot of shards from different indices.
What I observed is that all of them were marked red, whether they had
replicas or not. The global status was red.
After several hours the cluster did not recover.
I used the explain api and it told me about the missing primary shard of
the logs index. I then deleted the all the logs indices as they are non
essential.
Only then things started to improve and eventually the cluster recovered.
I conclude that the problem with non essential index was blocking cluster
recovery which is bad behavior.

This was a sterfull event which required fast action as production system
is down. So I can't really say if its a bug or able to reproduce this under
certain conditions.

**Comment by martijnvg** [comment]
@gary-harpaz What ES version is your production cluster running?

**Comment by martijnvg** [comment]
Also did you run the allocation explain on allocated shard for one of the logs index or the essential indices? I'm interested in what the output was of the allocation api for one of the essential indices.

**Comment by gary-harpaz** [comment]
6.0.1

I executed explain without specifying which index . The result complained about one of the logs indices.
After logs deletion during the recovery when I use explain on one of the essential indices it said:
"allocate_explanation": "cannot allocate because a previous copy of the primary shard existed but can no longer be found on the nodes in the cluster"
Which is a mystery to me. You see this error and you think it is final. But later this index did manage to recover.

**Comment by gary-harpaz** [comment]
Update on this issue.
After internal review of this incident we discovered that older essential indices were changed to have 0 replicas by retention process. We did this to save space. This explains why essential indices (older ones) were also status red when node crashed.

**Comment by martijnvg** [comment]
Thanks for the update and explanation.

I would suggest use the cat shards api to figure out which shards are not allocated and then use the allocation explain api to figure out why specific shards don't allocate automatically.

> "allocate_explanation": "cannot allocate because a previous copy of the primary shard existed but can no longer be found on the nodes in the cluster"

This means that a shard can not be restored, because the only existing copy of it is missing. Bringing back the node that had a copy of this shard, should fix this.

Also it turns out that there is already a way to give priority during shard recovery between indices.
Basically unallocated primaries are always allocated first then unallocated shards based on the `index.priority` index setting. This defaults to 1, but for your essential indices you can overwrite this.

Given the cluster health request can be send to check the status of specific indices then I think this issue can be closed?

**Comment by gary-harpaz** [comment]
No :)
I still would like those indices marked with 0 replicas to not trouble me if node crashes.
Don't want them to be marked red. Don't want global status to be red.

Does the "previous copy of a primary shard exitsted" error stay forever? Until I delete the index? How long does it stay?

**Comment by martijnvg** [comment]
We discussed issue and this is not a feature that Elasticsearch will support. To Elasticsearch, missing primary shards is a problematic situation that requires your attention, regardless of the index these shards belong to. Otherwise why do these indices exist in Elasticsearch to begin with? Maybe these indices are now not important anymore, but they were at some point in time. 

The [filtering support in the cluster health api](https://github.com/elastic/elasticsearch/issues/41367#issuecomment-484781586) should be sufficient to monitor the important indices in your cluster. In the case you want to continue to rely on the global status then you can think about writing a script that either [force allocates an empty primary shards](https://www.elastic.co/guide/en/elasticsearch/reference/6.0/cluster-reroute.html#_forced_allocation_on_unrecoverable_errors) or deletes the not important indices when the global status is red. This is something that Elasticsearch will never do automatically, because both actions involve deleting data, so writing a custom script is required.

> Does the "previous copy of a primary shard exitsted" error stay forever? Until I delete the index? How long does it stay?

That message will disappear when one of the following events occurs:
1) The primary shard was able to recover when the node that original had a copy re-joins the cluster.
2) An empty primary shard is forcefully allocated.
3) The index is deleted.
